File: day4/day4.ts
const fs = require('fs');
import { flatmap, map, pipe, sum, toarray } from 'powerseq';

function parseInput(input: string) {
  return pipe(
    input.split(/\n/),
    map(line => line.split("")),
    toarray()
  );
}

function inBounds(grid: string[][], row: number, col: number) {
  return row >= 0 && row < grid.length && col >= 0 && col < grid[0].length;
}

function checkWordInDirection(word, grid: string[][], row: number, col: number, dir: number[]): boolean {
  return Array.from({ length: word.length }, (_, i) => {
    const newRow = row + i * dir[0];
    const newCol = col + i * dir[1];
    return inBounds(grid, newRow, newCol) && grid[newRow][newCol] === word[i];
  }).every(Boolean);
}

//1.
function part1(grid: string[][]): number {
  const word = "XMAS";
  const directions = [[0, 1], [1, 0], [0, -1], [-1, 0], [1, 1], [1, -1], [-1, 1], [-1, -1]];
  return pipe(
    directions,
    flatmap(([dr, dc]) => pipe(
      grid,
      flatmap((row, r) =>
        row.map((_, c) => checkWordInDirection(word, grid, r, c, [dr, dc]) ? 1 : 0)
      )
    )),
    sum()
  )
}

//2.
function checkXShape(grid: string[][], row: number, col: number) {
  return grid[row][col] === 'A' &&
    (checkWordInDirection('AM', grid, row, col, [-1, -1]) && checkWordInDirection('AS', grid, row, col, [1, 1]) && (
      (checkWordInDirection('AM', grid, row, col, [-1, 1]) && checkWordInDirection('AS', grid, row, col, [1, -1])) ||
      (checkWordInDirection('AM', grid, row, col, [1, -1]) && checkWordInDirection('AS', grid, row, col, [-1, 1]))
    )) ||
    (checkWordInDirection('AM', grid, row, col, [1, 1]) && checkWordInDirection('AS', grid, row, col, [-1, -1]) && (
      (checkWordInDirection('AM', grid, row, col, [1, -1]) && checkWordInDirection('AS', grid, row, col, [-1, 1])) ||
      (checkWordInDirection('AM', grid, row, col, [-1, 1]) && checkWordInDirection('AS', grid, row, col, [1, -1]))
    ))
}

function part2(grid: string[][]): number {
  return pipe(
    grid,
    flatmap((row, r) =>
      row.map((_, c) => checkXShape(grid, r, c) ? 1 : 0)
    ),
    sum()
  )
}

const grid = parseInput(fs.readFileSync('./day4/input.txt', 'utf-8'))

console.log(part1(grid));
console.log(part2(grid));

File: day4/day4.go
package day4

import (
	"aoc2024/aochelpers"
	"strings"
)

func parseInput(input string) [][]string {
	lines := strings.Split(input, "\n")
	result := make([][]string, len(lines))
	for i, line := range lines {
			result[i] = strings.Split(line, "")
	}
	return result
}

func inBounds(grid [][]string, i , j int) bool {
	return i >= 0 && j >= 0 && i < len(grid) && j < len(grid[0])
}

func checkWordInDirection(word string, grid [][]string, row, col int, direction []int) bool {
	for i := 0; i < len(word); i++ {
		newRow := row + i * direction[0]
		newCol := col + i * direction[1]
		if !inBounds(grid, newRow, newCol) || grid[newRow][newCol] != string(word[i]) {
			return false
		}
	}
	return true
}

func Part1() {
	input := parseInput(aochelpers.ReadFileAsString("./day4/input.txt"))
	word := "XMAS"
	directions := [][]int{{1, 0}, {0, 1}, {1, 1}, {-1, -1}, {-1, 1}, {1, -1}, {0, -1}, {-1, 0}}
	result :=0
	for i := 0; i < len(input); i++ {
		for j := 0; j < len(input[0]); j++ {
			for _, direction := range directions {
				if checkWordInDirection(word, input, i, j, direction) {
					result++
				}
			}
		}
	}
	println("Day 4, part 1: ", result)
}

func checkXShape(grid [][]string, row, col int) bool {
	return grid[row][col] == "A" &&
		(checkWordInDirection("AM", grid, row, col, []int{-1, -1}) && checkWordInDirection("AS", grid, row, col, []int{1, 1}) && (
			(checkWordInDirection("AM", grid, row, col, []int{-1, 1}) && checkWordInDirection("AS", grid, row, col, []int{1, -1})) ||
			(checkWordInDirection("AM", grid, row, col, []int{1, -1}) && checkWordInDirection("AS", grid, row, col, []int{-1, 1})))) ||
		(checkWordInDirection("AM", grid, row, col, []int{1, 1}) && checkWordInDirection("AS", grid, row, col, []int{-1, -1}) && (
			(checkWordInDirection("AM", grid, row, col, []int{1, -1}) && checkWordInDirection("AS", grid, row, col, []int{-1, 1})) ||
			(checkWordInDirection("AM", grid, row, col, []int{-1, 1}) && checkWordInDirection("AS", grid, row, col, []int{1, -1}))))
}

func Part2() {	
	input := parseInput(aochelpers.ReadFileAsString("./day4/input.txt"))
	result := 0
	for i := 0; i < len(input); i++ {
		for j:= 0; j < len(input[0]); j++ {
			if checkXShape(input, i, j) {
				result++
			}
		}
	}
	println("Day 4, part 2: ", result)
}


